// ng-snotify/esm5/lib/components/buttons/buttons.component.js
import { __decorate } from "tslib";
import { ChangeDetectionStrategy, Component, Input, ViewEncapsulation } from '@angular/core';
import { SnotifyService } from '../../services/snotify.service';
var ButtonsComponent = /** @class */ (function () {
    function ButtonsComponent(service) {
        this.service = service;
    }
    /**
     * remove toast
     */
    ButtonsComponent.prototype.remove = function () {
        this.service.remove(this.toast.id);
    };
    ButtonsComponent.ctorParameters = function () { return [
        { type: SnotifyService }
    ]; };
    __decorate([
        Input()
    ], ButtonsComponent.prototype, "toast", void 0);
    ButtonsComponent = __decorate([
        Component({
            selector: 'ng-snotify-button',
            template: "<div class=\"snotifyToast__buttons\">\r\n  <button\r\n    type=\"button\"\r\n    *ngFor=\"let button of toast.config.buttons\"\r\n    [ngClass]=\"{ 'snotifyToast__buttons--bold': button.bold }\"\r\n    (click)=\"button.action ? button.action(toast) : remove()\"\r\n  >\r\n    {{ button.text }}\r\n  </button>\r\n</div>\r\n",
            changeDetection: ChangeDetectionStrategy.OnPush,
            encapsulation: ViewEncapsulation.None
        })
        /**
         * Buttons component
         */
    ], ButtonsComponent);
    return ButtonsComponent;
}());
export { ButtonsComponent };
