// ng-snotify/esm5/lib/snotify.module.js
import { __decorate } from "tslib";
import { NgModule } from '@angular/core';
import { SnotifyComponent } from './components/snotify/snotify.component';
import { SnotifyService } from './services/snotify.service';
import { KeysPipe } from './pipes/keys.pipe';
import { TruncatePipe } from './pipes/truncate.pipe';
import { CommonModule } from '@angular/common';
import { ButtonsComponent } from './components/buttons/buttons.component';
import { PromptComponent } from './components/prompt/prompt.component';
import { ToastComponent } from './components/toast/toast.component';
var SnotifyModule = /** @class */ (function () {
    function SnotifyModule() {
    }
    SnotifyModule_1 = SnotifyModule;
    SnotifyModule.forRoot = function () {
        return {
            ngModule: SnotifyModule_1,
            providers: [SnotifyService]
        };
    };
    var SnotifyModule_1;
    SnotifyModule = SnotifyModule_1 = __decorate([
        NgModule({
            imports: [CommonModule],
            declarations: [SnotifyComponent, ToastComponent, TruncatePipe, ButtonsComponent, PromptComponent, KeysPipe],
            exports: [SnotifyComponent, TruncatePipe, KeysPipe]
        })
    ], SnotifyModule);
    return SnotifyModule;
}());
export { SnotifyModule };
